tests/0151-purge-brokers.c

```diff
@@ -0,0 +1,410 @@
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "test.h"
+
+static rd_atomic32_t do_test_remove_then_add_received_terminate;
+static rd_atomic32_t verification_complete;
+
+/** @brief Verify that \p expected_broker_ids
+ *         and \p actual_broker_ids correspond in
+ *         count and value. Sorts \p actual_broker_ids .
+ */
+static rd_bool_t fetch_metadata_verify_brokers(int32_t *expected_broker_ids,
+                                               size_t expected_broker_id_cnt,
+                                               int32_t *actual_broker_ids,
+                                               size_t actual_broker_id_cnt) {
+        size_t i;
+        if (actual_broker_id_cnt != expected_broker_id_cnt)
+                return rd_false;
+
+        for (i = 0; i < actual_broker_id_cnt; i++) {
+                if (actual_broker_ids[i] != expected_broker_ids[i])
+                        return rd_false;
+        }
+        return rd_true;
+}
+
+static void fetch_metadata(rd_kafka_t *rk,
+                           int32_t *expected_broker_ids,
+                           size_t expected_broker_id_cnt,
+                           rd_bool_t (*await_after_action_cb)(int action),
+                           int action) {
+        const rd_kafka_metadata_t *md = NULL;
+        rd_kafka_resp_err_t err;
+        size_t actual_broker_id_cnt = 0;
+        int32_t *actual_broker_ids  = NULL;
+        size_t i;
+        int timeout_usecs                      = 10000000;
+        int64_t abs_timeout_us                 = test_clock() + timeout_usecs;
+        rd_bool_t continue_requesting_metadata = rd_true;
+
+        TEST_SAY("Waiting for up to 10s for metadata update\n");
+
+        /* Trigger Metadata request which will update learned brokers. */
+        do {
+                err = rd_kafka_metadata(rk, 0, NULL, &md, tmout_multip(5000));
+                if (md) {
+                        rd_kafka_metadata_destroy(md);
+                        md = NULL;
+                } else if (err != RD_KAFKA_RESP_ERR__TRANSPORT)
+                        TEST_ASSERT(!err, "%s", rd_kafka_err2str(err));
+
+                rd_usleep(100 * 1000, 0);
+
+                RD_IF_FREE(actual_broker_ids, rd_free);
+                actual_broker_ids =
+                    rd_kafka_brokers_learned_ids(rk, &actual_broker_id_cnt);
+                continue_requesting_metadata = test_clock() <= abs_timeout_us;
+
+                continue_requesting_metadata &= !fetch_metadata_verify_brokers(
```

Comment:
Use logical AND operator instead of bitwise AND operator. Though the result is going to be same here.